src/librustc/session/config.rs

```diff
@@ -933,7 +957,28 @@ pub fn build_session_options(matches: &getopts::Matches) -> Options {
         output_types.insert(OutputType::Exe, None);
     }
 
-    let cg = build_codegen_options(matches, color);
+    let mut cg = build_codegen_options(matches, color);
+
+    // Issue #30063: if user requests llvm-related output to one
+    // particular path, disable codegen-units.
+    if matches.opt_present("o") && cg.codegen_units != 1 {
```

Comment:
I'd personally prefer to omit the check here for `-o` and always execute this logic if `codegen_units != 1`, that way we maintain equivalence between `rustc -o foo foo.rs` and `rustc foo.rs` (for all emit types)